include/vapor/ArbitrarilyOrientedRegularGrid.h

```diff
@@ -0,0 +1,103 @@
+#pragma once
+
+#include <ostream>
+#include <vector>
+#include <vapor/common.h>
+#include <vapor/RegularGrid.h>
+#include <glm/glm.hpp>
+
+// clang-format off
+
+struct planeDescription {
+    size_t sideSize;
+    std::vector<double> origin;
+    std::vector<double> rotation;
+    VAPoR::CoordType boxMin;
+    VAPoR::CoordType boxMax;
+    VAPoR::CoordType domainMin;
+    VAPoR::CoordType domainMax;
+};
+
+namespace VAPoR {
+
+//! Create a 2D RegularGrid that is rotated along an arbitrary orientation.  This new grid
+//! samples a given 3D Grid object along a plane, through the entire domain.  The "planeDescription" 
+//! variable includes parameters that define the 2D plane's origin, rotation, and valid extents.  Points on the
+//! 2D RegularGrid that are outside of the valid extents will be assigned as
+//! missingValues.
+//!
+//! Note: The boxMin and boxMax values within the planeDescription should come from
+//! the current RenderParams' Box class.  The domainMin and domainMax values should
+//! come from the DataMgr::GetVariableExtents function.  Taking these values from
+//! the given 3D grid can produce erroneous results when the 3D grid's region is reduced.
+//!
+//! \param[in] grid3d A variable's 3D grid to be sampled along a plane, to generate a 2D grid object
+//! \param[in] description A set of std::vector<dobule> values that define the origin, rotation, and extents of the returned 2D grid
+//! \param[in] dims The sampling rate that the new grid will be defined upon, along its X and Y axes
+//! \param[out] data An array of sideSize*sideSize values that contain floating point values of the sampled data.
+//!
+
+class VDF_API ArbitrarilyOrientedRegularGrid : public RegularGrid {
+public:
+    ArbitrarilyOrientedRegularGrid(
+        const VAPoR::Grid* grid3d,
+        planeDescription& pd,
+        const DimsType& dims,
+        std::shared_ptr<float>& blks
+    );
+
+    ArbitrarilyOrientedRegularGrid() = default;
+    virtual ~ArbitrarilyOrientedRegularGrid() = default;
+
+    static std::string GetClassType() { return ("ArbitrarilyOrientedRegular"); }
+    std::string        GetType() const override { return (GetClassType()); }
+
+    //! \copydoc Grid::GetUserCoordinates()
+    //
+    virtual void GetUserCoordinates(const DimsType &indices, CoordType &coords) const override;
+
+    //! Retrieve the winding order that pertains to the generated array of data, if it is
+    //! to be drawn as a texture as is done in the SliceRenderer's vertex buffer.  The winding order is a series
+    //! of six 3D vertices that describe two triangles that define how the texture is drawn in space.
+    void GetWindingOrder( std::vector<double>& windingOrder ) const;
```

Comment:
The "winding order" and "two triangles" confuse me. Does this class define a plane, or two adjacent triangles?

Reply:
It describes both.

The [texture coordinate buffer](https://github.com/NCAR/VAPOR/blob/dd7ac4ee9d3aee8e508b6a40723e83c548ef9337/lib/render/SliceRenderer.cpp#L101) needs the four coordinates that describe the corners of the plane.

We need to draw the plane as a pair of triangles.  [The vertex buffer](https://github.com/NCAR/VAPOR/blob/dd7ac4ee9d3aee8e508b6a40723e83c548ef9337/lib/render/SliceRenderer.cpp#L112) needs to know the winding order.

Reply:
Still confused. Polygon winding orders that I'm familiar with are typically expressed as rotations about a normal: clockwise, or counter-clockwise (CCW), not as a list of points. So maybe that is part of my confusion. Moreover, the convention for 2D Grids is that the winding order is CCW. 2D Grids are always constructed this way. Since you are constructing the grid from scratch, don't you have the ability to use a consistent winding order?

Reply:
> Polygon winding orders that I'm familiar with are typically expressed as rotations about a normal: clockwise, or counter-clockwise (CCW), not as a list of points.

To whoever's reading this, John and I met about this last Friday.  I'm hoping I can absolve the confusion here.

The winding order does more than list the vertices of the texture according to the right-hand-rule (counter clockwise with reference to the normal).  

It lists the order of the vertices needed to draw the two triangles that compose the texture, used the [the vertex buffer](https://github.com/NCAR/VAPOR/blob/dd7ac4ee9d3aee8e508b6a40723e83c548ef9337/lib/render/SliceRenderer.cpp#L112).

@clyne - As we agreed I've moved the winding order into the Renderer library.  It now exists in SliceRenderer.cpp:211 (sorry I can't copy a perma-link for some reason).  I think this requires more discussion, because any other renderer that will need to draw a texture will need to duplicate this same ordering.